pwndbg/gdblib/functions.py

```diff
@@ -136,6 +136,134 @@ def base(name_pattern: gdb.Value | str):
+
+    Example:
+    ```
+    pwndbg> vmmap
+    0x555555554000     0x555555555000 r--p     1000       0 p1
+    0x555555555000     0x555555556000 r-xp     1000    1000 p1
+    0x555555556000     0x555555557000 r--p     1000    2000 p1
+    0x555555557000     0x555555558000 r--p     1000    2000 p1
+    0x555555558000     0x555555559000 rw-p     1000    3000 p1
+    0x555555559000     0x55555557a000 rw-p    21000       0 [heap]
+    [...]
+    pwndbg> p/x $heap()
+    $1 = 0x555555559000
+    pwndbg> p/x $heap(0x10)
+    $2 = 0x555555559010
+    pwndbg> x/32bx $heap()
+    0x555555559000:	0x00	0x00	0x00	0x00	0x00	0x00	0x00	0x00
+    0x555555559008:	0x91	0x02	0x00	0x00	0x00	0x00	0x00	0x00
+    0x555555559010:	0x00	0x00	0x00	0x00	0x00	0x00	0x00	0x00
+    0x555555559018:	0x00	0x00	0x00	0x00	0x00	0x00	0x00	0x00
+    ```
+    """
+    base_addr = gdb.parse_and_eval('$base("heap")')
+    return int(base_addr) + int(offset)
+
+
+@GdbFunction(only_when_running=True)
+def stack(offset: gdb.Value = gdb.Value(0)) -> int:
+    """
+    Return the base address of the stack mapping, optional offset allowed.
+
+    Example:
+    ```
+    pwndbg> vmmap
+    [...]
+    0x7ffff7fc5000     0x7ffff7fc6000 r--p     1000       0 /usr/lib/x86_64-linux-gnu/ld-linux-x86-64.so.2
+    0x7ffff7fc6000     0x7ffff7ff1000 r-xp    2b000    1000 /usr/lib/x86_64-linux-gnu/ld-linux-x86-64.so.2
+    0x7ffff7ff1000     0x7ffff7ffb000 r--p     a000   2c000 /usr/lib/x86_64-linux-gnu/ld-linux-x86-64.so.2
+    0x7ffff7ffb000     0x7ffff7ffd000 r--p     2000   36000 /usr/lib/x86_64-linux-gnu/ld-linux-x86-64.so.2
+    0x7ffff7ffd000     0x7ffff7fff000 rw-p     2000   38000 /usr/lib/x86_64-linux-gnu/ld-linux-x86-64.so.2
+    0x7ffffffde000     0x7ffffffff000 rw-p    21000       0 [stack]
+    [...]
+    pwndbg> p/x $stack()
+    $1 = 0x7ffffffde000
+    pwndbg> x/gx $stack()
+    0x7ffffffde000:	0x0000000000000000
+    ```
+    """
+    base_addr = gdb.parse_and_eval('$base("stack")')
+    return int(base_addr) + int(offset)
+
+
+def _resolve_symbol_address(candidates: list[str]) -> int | None:
+    """
+    Try to resolve any of the given symbol names to an int address.
+    Returns the address, or None if none of them can be resolved.
+    """
+    for name in candidates:
+        for expr in (name, f"&{name}"):
+            try:
+                val = gdb.parse_and_eval(expr)
+            except gdb.error:
+                continue
+            return int(val)
+
+    return None
+
+
+@GdbFunction(only_when_running=True)
+def bss(offset: gdb.Value = gdb.Value(0)) -> int:
+    """
+    Return the address of the program's BSS start. Optional offset allowed.
+
+    Example:
+    ```
+    pwndbg> p/x $bss()
+    $1 = 0x555555558010
+    pwndbg> p/x $bss(0x10)
+    $2 = 0x555555558020
+    ```
+    """
+    bss_symbols = [
+        "__bss_start",
+        "_bss_start",
+        "__bss_start__",
+        "bss_start",
+        "_edata",
```

Comment:
I wonder if there is a way to get this based on `.bss` segment/section name but idk